aten/src/ATen/core/typeid.h

```diff
@@ -81,51 +90,145 @@ AT_DEFINE_HASH_FOR_IDWRAPPER(caffe2::TypeIdentifier)
-               " FULL_CAFFE2=1 (if you build latest master, the ONNX scripts are "
-               "updated to do this for you.) "
-               "For more context, see https://github.com/pytorch/pytorch/issues/10460");
-    }
-    gRegisteredTypeNames().insert(name);
-    gTypeNames()[id] = name;
-#else // __GXX_RTTI
-    if (literal_name.empty()) {
-      gTypeNames()[id] = "(RTTI disabled, cannot show name)";
-    } else {
-      gTypeNames()[id] = literal_name;
-    }
-#endif // __GXX_RTTI
+inline void _Ctor(void* ptr, size_t n) {
+  T* typed_ptr = static_cast<T*>(ptr);
+  for (size_t i = 0; i < n; ++i) {
+    new (typed_ptr + i) T;
   }
+}
+
+template <typename T>
+inline void _CtorNotDefault(void* /*ptr*/, size_t /*n*/) {
+  _ThrowRuntimeTypeLogicError(
+      "Type " + std::string(at::demangle_type<T>()) +
+      " is not default-constructible.");
+}
+
+template <
+    typename T,
+    typename std::enable_if<std::is_default_constructible<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::PlacementNew* _PickCtor() {
+  return &_Ctor<T>;
+}
+
+template <
+    typename T,
+    typename std::enable_if<!std::is_default_constructible<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::PlacementNew* _PickCtor() {
+  return &_CtorNotDefault<T>;
+}
+
+/**
+ * Typed copy function for classes.
+ */
+template <typename T>
+inline void _Copy(const void* src, void* dst, size_t n) {
+  const T* typed_src = static_cast<const T*>(src);
+  T* typed_dst = static_cast<T*>(dst);
+  for (size_t i = 0; i < n; ++i) {
+    typed_dst[i] = typed_src[i];
+  }
+}
+
+/**
+ * A placeholder function for types that do not allow assignment.
+ */
+template <typename T>
+inline void _CopyNotAllowed(
+    const void* /*src*/,
+    void* /*dst*/,
+    size_t /*n*/) {
+  _ThrowRuntimeTypeLogicError(
+      "Type " + std::string(at::demangle_type<T>()) +
+      " does not allow assignment.");
+}
+
+template <
+    typename T,
+    typename std::enable_if<std::is_copy_assignable<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::TypedCopy* _PickCopy() {
+  return &_Copy<T>;
+}
+
+template <
+    typename T,
+    typename std::enable_if<!std::is_copy_assignable<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::TypedCopy* _PickCopy() {
+  return &_CopyNotAllowed<T>;
+}
+
+/**
+ * Destructor for non-fundamental types.
+ */
+template <typename T>
+inline void _Dtor(void* ptr, size_t n) {
+  T* typed_ptr = static_cast<T*>(ptr);
+  for (size_t i = 0; i < n; ++i) {
+    typed_ptr[i].~T();
+  }
+}
+
+template<class T> const char* _TypeName() noexcept;
+
+template<class T, class Enable = void>
+struct TypeMetaDataRegistry final {
+  static_assert(!std::is_same<T, T>::value, "This should never be picked since TypeMetaDataRegistry has specialisations for all cases");
 };
 
+template <typename T>
+struct TypeMetaDataRegistry<T, c10::guts::enable_if_t<std::is_fundamental<T>::value || std::is_pointer<T>::value>> final {
```

Comment:
so.... will it work as expected across binary boundaries?

the reason current type meta was setup this way with CAFFE_KNOWN_TYPE explicit anntotations is that static variables in templates are prone to double-declarations when shared libraries are loaded with RTLD_GLOBAL.

it might be fine as long as we don't use TypeMetaDataRegistry pointer for comparison, but it looks like you do above in operator==

The general safety rule is to avoid static variables in templates or in inline functions in .h files all together. Thus you can move this initialization to the constructor and explicitly specialize/instantiate TypeMetaDataRegistry only within CAFFE_KNOWN_TYPE. Would that work?

Reply:
true, I'll switch it back to id based comparison.